### DISCUSSION TITLE
loosing withcount

### DISCUSSION THREAD

**Discussion by ChrisSantiago82** [livewire/livewire#2950] [1. Help]
I'm having an index page build with livewire. I want to show the count of a relationship and I'm using the withcount from eloquent. On the first render it works fine and afterward that field/data gets lost.

What would be the prober way for me to do this?

**Comment by drbornot** [comment]
Maybe you could show us how you're handling that on your code

**Comment by ChrisSantiago82** [comment]
Hey,
actually there is not that much code to show.

I have a query that I do in the mount method:

$this->tournaments= Tournament::withCount('players')->get();

Then on the tournament collection I have the players_count variable:
@foreach($tournaments as $tournament)
{{ tournament->players_count }}
@endforeach

It's based on this: https://laravel.com/docs/8.x/eloquent-relationships#counting-related-models

The idea is that I don't want to load the players models, I just need the count of them for each tournament.
First time when the page is rendered that field on the collection is there. When livewire makes another roundtrip, like when a button is pressed or an input field is being changed, that field is lost in the collection.

**Comment by ChrisSantiago82** [comment]
Nobody else has this issue? Nobody uses withcount in a query?

**Reply by drbornot** [reply]
Create an external function that you can call on render, instead doing it in mount like
```
public function loadPlayersCount()
{
   return Tournament::withCount('players')->get();
}
public function render()
{
   $this->tournaments = $this->loadPlayersCount(); // -> or you can call it after any action you want, like reload the property
   return view('your-blade');
}
```

**Comment by ChrisSantiago82** [comment]
Yes, this could work thanks. 
The only problem is that the query would run on every render. For example if that component has an input field, then on almost every key stroke that query would run.

**Reply by drbornot** [reply]
For it I add the "// -> or you can call" comment, you can detach it from render function and call the reload on any other function after an action that need refresh the property

**Reply by ChrisSantiago82** [reply]
Yes, but the problem is that it would loose the count data by every render, for example every key stroke if it's an input field.
Either it's going to be heavy on the database or I'm loosing data on the screen. There should be another option.

**Comment by ryatkins** [comment]
@ChrisSantiago82 this looks like a good use case for computed properties.
https://laravel-livewire.com/docs/2.x/computed-properties

Remove `public $tournaments` as a class property as the computed property will create it itself.

```
public function getTournamentsProperty() {

    return Tournament::withCount('players')->get();

}
```

Reference as $this->tournaments in your view
```
@foreach($this->tournaments as $tournament)
      {{ tournament->players_count }}
@Endforeach
```

**Comment by ChrisSantiago82** [comment]
@ryatkins, That looks very interesting. I didn't know about the computed-properties in livewire.
But would the result not be the same if I put the query in the render method? Wouldn't that just make that the query runs every time when something updates like an input field?

**Comment by ryatkins** [comment]
It does some level of caching for a single request cycle, but if you're doing a roundtrip to the server on every keyboard input it might run this query again.

On your input fields I would add `wire:model.defer` so that round trips to the server only happen on a click or submit. 

> Computed properties are cached for an individual Livewire request lifecycle. Meaning, if you call `$this->post` 5 times in a component's blade view, it won't make a separate database query every time.

**Comment by ChrisSantiago82** [comment]
I saw that it would do it once for one request. But we are sending all the collection around with all the data, and just because of the count field, to run the whole query again and again, seems a little bit too much.

It seems that the livewire looses all the fields that are not fields in the table. If you wire a model directly you need to add that field to the rule or it looses the data as well.
Wouldn't it be cool if livewire wouldn't loose that data? The count is a very nice Laravel feature. And there are others, like withSum.  Maybe there could be a way to tell livewire that it should keep that data? I don't know much about livewire, where could I implement this function? And is that something that would be published?

**Comment by ryatkins** [comment]
@ChrisSantiago82 

Another option is to split your component into two components, one for your count and one for your data.

**Comment by joshhanley** [comment]
@ChrisSantiago82 the reason `withCount` isn't persisted is because Livewire has no way of knowing that a custom query has been run. 

Currently, the way Livewire works, is if you add an eloquent collection to a public property, the only pieces of information Livewire keeps track of by default and sends to the front end, is the class of the models in the collection and a list of their IDs, this is to keep the payload small. It is also more secure by not just randomly sending all model data to the front end.

When a request is made from the front end, the model class and the IDs are sent to the backend, where Livewire queries the database for those models and then repopulates the collection.

This is why custom queries like `withCount` doesn't work with Livewire, as it has no way of knowing that was called on the collection in the first place, and as such has no way of knowing to call it again when the models are rehydrated from the database.

So there are a couple of solutions as mentioned above. 

One, just run the query in your render method (or extract to a method but call it in render) and pass the collection into the view as suggested by @drbornot .

Two, use a computed property as suggested by @ryatkins.

Three, you could make use of a custom hydrate method where you can place your query to ensure it get's run every time the collection is hydrated, check out `hydrateFoo` example here https://laravel-livewire.com/docs/2.x/lifecycle-hooks.

But, no matter which way you do it, there will always be a database query run for every Livewire request, as that is the nature of PHP. If you want the full eloquent models to be available in the class, they need to be pulled from the database.

If you are concerned about database usage, then I'd recommend converting your eloquent collection to an array and then storing that in your public property. But that then increases the payload sent between the backend and the front end, as it needs to transmit the full array in both directions to ensure the property can be hydrated with the correct data.

Personally, I don't see any issue with the database queries running, but putting things like `.defer` or `.lazy`, or even a longer `.debounce` on your inputs is recommended to minimise the network requests being sent, as out of the two, generally the slower part is the network, not the DB queries (if your DB is indexed properly).

Hope this helps, let me know if you want me explain anything else 🙂

**Comment by ChrisSantiago82** [comment]
I thank you all for your good advices.

I think the array solution is an interesting one, but probably I would just make an array with the count data to keep it slim.

I understand how livewire works. I opened this discussion because lately I see more and more bad from people who use livewire and the programmers don't even know what they are doing because it's just working.

For example, in this case, the programmer eager loaded all players and counted them with the ->count() method from the collection. That means there were so much information being send to the browser to get the count. And on the second round trip, the eager loading didn't work anymore and on each round trip there were 800 queries running.

I'm not saying that all data should be persist, but wouldn't it be nice if there would be a way to tell livewire that it should keep that field (for example withcount) or at lest tell livewire that there is a withcount that should be loaded again. There is a method like this for loading the relationships again, but that method didn't even make it to the livewire documents. My question is, shouldn't it be simpler to produce good code?

**Reply by joshhanley** [reply]
@ChrisSantiago82 ouch 🙈 that's a lot of queries.

If someone could come up with a clean way to tell Livewire that there is a `withCount` used, I'm sure a PR would be welcome.

I can't see an easy and clean way to do it other than using the suggestions above. Plus Livewire currently uses the same data serialiser that Laravel uses for queued jobs, so you'd have the same problem in queued jobs too.

But also, is that the responsibility of Livewire to make sure developers are writing good code? You can have the same issues with Laravel if you're not careful and monitoring your queries.

Can you expand on the method for loading relationships that you are talking about that's not documented? As if something has been missed then maybe we can update the documentation for it.

Definitely worth discussing though 🙂

**Comment by bram-pkg** [comment]
I am also running into this issue, where calling `withCount` on a query which is stored in a `Collection` class property, loses the count information after a refresh.

Code sample:

```php
public string $course = '';
/** @var Collection<Course> */
public Collection $courses;

public function mount()
{
    $this->courses = Course::query()
        ->withCount('students')
        ->orderBy('name')
        ->get();
}
```

```html
<input type="search" wire:model="course">

<select wire:model="course">
  <option value="" selected>Choose course</option>
  @foreach($courses as $course)
    <option value="{{ $course->id }}">
      {{ $course->name }} ({{ $course->students_count }})
    </option>
  @endforeach
</select>
```

This causes the `$course->students_count` data to be lost after typing in the `search` field.

**Reply by joshhanley** [reply]
@bram-pkg please see my explanation above as to why `->withCount()` won't work and some possible work arounds 🙂 https://github.com/livewire/livewire/discussions/2950#discussioncomment-1208795

**Reply by ChrisSantiago82** [reply]
On the relationship we did loadmissing (I don't think it's documented).
Maybe we could do something like countmissing?

**Reply by joshhanley** [reply]
@ChrisSantiago82 does Laravel have a `countMissing` method? That'd be handy if it did for this scenario!

**Reply by ChrisSantiago82** [reply]
There is a loadCount:

https://laravel.com/docs/8.x/eloquent-relationships#deferred-count-loading

**Reply by joshhanley** [reply]
@ChrisSantiago82 ah yeah cool, so you could put that in a hydrate method, like

```php

public $users;

public hydrateUsers()
{
    $this->users->loadCount('roles');
}
```

Something like that should work.

**Reply by ChrisSantiago82** [reply]
I agree.
It seems that other people are having the same problem.
How can we make sure they don't get stuck at the same issue? Maybe we could add it to the documentation? What do you think?

**Reply by joshhanley** [reply]
@ChrisSantiago82 yeah that is a tough one, as there is a lot of possible variations of what is/isn't supported in terms of querying models and then saving them to a public property (ie custom selects, joins, etc, don't work, but filters and orderby and stuff does).

So I'm not sure how best to document something like that. But if you have any ideas PR's are always welcome to the docs 🙂
https://github.com/livewire/docs

**Reply by Spriithy** [reply]
I think adding the ways to circumvent this problem to the docs would be nice since Google doesn't index this discussion properly when using the right keywords ^^"

**Reply by joshhanley** [reply]
@Spriithy yeah agreed. As above though, I'm not sure how best to document it. If you were searching for it, what would you be searching for, and where do you think that information should be in the docs so it's easily found?

**Reply by Spriithy** [reply]
@joshhanley  Well, I thought I had replied by mail (as one can do with issues) but it seems it didn't work ! So I'm pasting my initial reply here.

My general understanding of the issue, out of what I have read online, is that people tend to overlook the fact that Livewire doesn't bundle every mething from the models forward to the client.

I think insisting on that point and hinting towards the hydrate/loadMissing combination is the way to go in the docs.

Key terms might look like: missing with and withCount data after component refresh or something like that.

Hope this helped!

**Reply by joshhanley** [reply]
@Spriithy ah interesting, I think it did work as I read your reply the other day, maybe it disappeared? 🤔

Yeah that makes sense, thanks!

**Comment by bram-pkg** [comment]
I have now solved it by adding a `hydrate()` method to my Livewire component that calls `loadCount()` on the relationship that needs to be loaded.

The "normal" data is loaded by the `mount()` method, using `Course::query()->withCount('students')->orderBy('name')->get()`.

![image](https://user-images.githubusercontent.com/13304739/147385948-12a80c04-6cc1-48d6-a1e5-3e48ba8ae43f.png)

**Reply by dasundev** [reply]
@bram-pkg thanks man!

**Comment by ercanertan** [comment]
Using loadSum() and loadCount() in **rendering()** method does the job.
